File: cpp/headless_model.cpp
#include "basicsim/jaamsim_model.h"
#include "basicsim/simulation.h"
#include "events/event_manager.h"
#include "json/json_parser.h"
#include "json/json_writer.h"
#include "remote/zmq_bridge.h"

#include <iostream>
#include <fstream>
#include <sstream>

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <config-file>\n";
        return 1;
    }

    try {
        jaamsim::basicsim::JaamSimModel model;

        model.auto_load();

        std::ifstream config_file(argv[1]);
        if (!config_file) {
            std::cerr << "Cannot open config file: " << argv[1] << '\n';
            return 1;
        }
        std::stringstream buffer;
        buffer << config_file.rdbuf();
        std::string config_text = buffer.str();

        auto config = jaamsim::json::parse_json(config_text);
        model.configure(config);

        jaamsim::remote::ZmqBridge bridge;
        bridge.start(model);

        std::cout << "ZMQ bridge started on tcp://*:5556 (PUB) and tcp://*:5557 (ROUTER)\n";
        std::cout << "Running simulation...\n";

        auto run_result = model.start();

        std::cout << "Simulation ended. Sim time: " << run_result.sim_time << " s\n";

        bridge.stop();
    }
    catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << '\n';
        return 1;
    }

    return 0;
}

File: src/main/java/com/jaamsim/remote/HeadlessModel.java
/*
 * JaamSim Discrete Event Simulation
 * Copyright (C) 2026 JaamSim Software Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.jaamsim.remote;

import java.io.File;

import com.jaamsim.basicsim.JaamSimModel;
import com.jaamsim.basicsim.RunListener;

public class HeadlessModel {

	public static void main(String[] args) {
		if (args.length < 1) {
			System.err.println("Usage: HeadlessModel <config-file>");
			System.exit(1);
		}

		File configFile = new File(args[0]);
		if (!configFile.exists()) {
			System.err.println("Config file not found: " + args[0]);
			System.exit(1);
		}

		JaamSimModel simModel = new JaamSimModel();
		simModel.autoLoad();

		try {
			simModel.configure(configFile);
		} catch (Exception e) {
			System.err.println("Configure warnings: " + e.getMessage());
		}

		ZmqBridge bridge = new ZmqBridge();
		bridge.start(simModel);

		System.out.println("ZMQ bridge started on tcp://*:5556 (PUB) and tcp://*:5557 (ROUTER)");
		System.out.println("Running simulation...");

		final Object lock = new Object();
		simModel.start(new RunListener() {
			@Override
			public void runEnded() {
				synchronized (lock) { lock.notifyAll(); }
			}

			@Override
			public void handleRuntimeError(JaamSimModel sm, Throwable t) {
				System.err.println("Runtime error: " + t.getMessage());
				t.printStackTrace();
				synchronized (lock) { lock.notifyAll(); }
			}
		});

		synchronized (lock) {
			try { lock.wait(); } catch (InterruptedException e) {}
		}

		System.out.println("Simulation ended. Sim time: " + simModel.getSimTime() + " s");
		bridge.stop();
		System.exit(0);
	}

}
